### DISCUSSION TITLE
What does the 'start' function on output.alsa do?

### DISCUSSION THREAD

**Discussion by vrtisworks** [savonet/liquidsoap#2290] [Q&A]
Here is some simple code that I don't understand.
`#!/home/liquidsoap/.opam/default/bin/liquidsoap
alsaDevice = "hw:CARD=Headphones"
thesource=single("testSong.wav")
localOut = output.alsa(id='localAudio',
	device="#{alsaDevice}",
	start=true,
	fallible=true,thesource)
log("localOut: #{localOut.is_started()}:#{localOut.id()}")
localOut.start()
log("localOut: #{localOut.is_started()}:#{localOut.id()}")
`
This works fine.
But if I set the 'start=' parameter to start=false, nothing ever plays.

The log shows that 'is_started' goes from 'false' to 'true'

What concept am I missing?

Thanks.

**Comment by toots** [comment]
It is a legacy issue. We have two alsa output implementations, one when `bufferize=true` and one when `bufferize=false`. Only the implementation with `bufferize=false` seems to have a `start`/`stop` method implemented and `bufferize=true` is the default.. This should be cleaned up: https://github.com/savonet/liquidsoap/issues/2291

**Reply by vrtisworks** [reply]
I tried it with bufferize=false, start=true - I do get music, but lots off buffer under runs

`
liquidsoap@pithree:~/audiobox $ ./justTesting.sh
2022/03/14 16:26:11 >>> LOG START
2022/03/14 16:26:09 [main:3] Liquidsoap 2.0.3-1
2022/03/14 16:26:09 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.6 sedlex=2.5 menhirLib=20211128 curl=0.9.2 dtools=0.4.4 duppy=0.9.2 cry=0.6.6 mm=0.7.4 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 mad=0.5.2 dynlink=[distributed with Ocaml] lame=0.3.5 alsa=0.3.0 taglib=0.3.9 camomile=1.0.2
2022/03/14 16:26:09 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2022/03/14 16:26:09 [clock:3] Using builtin (low-precision) implementation for latency control
2022/03/14 16:26:11 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2022/03/14 16:26:11 [frame:3] Video frame size set to: 1280x720
2022/03/14 16:26:11 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2022/03/14 16:26:11 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2022/03/14 16:26:11 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2022/03/14 16:26:11 [lang:3] localOut: false:localAudio
2022/03/14 16:26:11 [source:3] Using ALSA 1.2.4.
2022/03/14 16:26:11 [source:3] Falling back on interleaved S16LE
2022/03/14 16:26:11 [source:3] Samplefreq=44100Hz, Bufsize=1764B, Frame=4B, Periods=4
2022/03/14 16:26:11 [lang:3] localOut: true:localAudio
2022/03/14 16:26:11 [sandbox:3] Sandboxing disabled
2022/03/14 16:26:11 [video.converter:3] Couldn't find preferred video converter: ffmpeg.
2022/03/14 16:26:11 [audio.converter:3] Using samplerate converter: native.
2022/03/14 16:26:11 [single_0:3] testSong.wav is static, resolving once for all...
2022/03/14 16:26:11 [clock.alsa:3] Streaming loop starts in auto-sync mode
2022/03/14 16:26:11 [testSong.wav:3] Prepared "testSong.wav" (RID 0).
2022/03/14 16:26:11 [clock.alsa:3] Delegating synchronisation to active sources
2022/03/14 16:26:13 [localAudio:2] Underrun! You may minimize them by increasing the buffer size.
2022/03/14 16:26:13 [localAudio:2] Trying to recover..
2022/03/14 16:26:13 [localAudio:2] Underrun! You may minimize them by increasing the buffer size.
2022/03/14 16:26:13 [localAudio:2] Trying to recover..
2022/03/14 16:26:15 [localAudio:2] Underrun! You may minimize them by increasing the buffer size.
2022/03/14 16:26:15 [localAudio:2] Trying to recover..
^C2022/03/14 16:26:15 [main:3] Shutdown started!
2022/03/14 16:26:15 [main:3] Waiting for main threads to terminate...
2022/03/14 16:26:15 [clock.alsa:3] Delegating synchronisation to CPU clock
2022/03/14 16:26:15 [clock.alsa:3] Streaming loop stopped.
2022/03/14 16:26:15 [main:3] Main threads terminated.
2022/03/14 16:26:15 [threads:3] Shutting down scheduler...
2022/03/14 16:26:15 [threads:3] Scheduler shut down.
2022/03/14 16:26:15 [main:3] Cleaning downloaded files...
2022/03/14 16:26:15 [main:3] Freeing memory...
liquidsoap@pithree:~/audiobox $

`

**Reply by vrtisworks** [reply]
If I use bufferize=false, start=false, - I do not get music   Nothing appears to happen. Here is the log from that.

liquidsoap@pithree:~/audiobox $ ./justTesting.sh
2022/03/14 16:45:45 >>> LOG START
2022/03/14 16:45:44 [main:3] Liquidsoap 2.0.3-1
2022/03/14 16:45:44 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.6 sedlex=2.5 menhirLib=20211128 curl=0.9.2 dtools=0.4.4 duppy=0.9.2 cry=0.6.6 mm=0.7.4 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 mad=0.5.2 dynlink=[distributed with Ocaml] lame=0.3.5 alsa=0.3.0 taglib=0.3.9 camomile=1.0.2
2022/03/14 16:45:44 [dynamic.loader:3] Could not find dynamic module for fdkaac encoder.
2022/03/14 16:45:44 [clock:3] Using builtin (low-precision) implementation for latency control
2022/03/14 16:45:45 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2022/03/14 16:45:45 [frame:3] Video frame size set to: 1280x720
2022/03/14 16:45:45 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2022/03/14 16:45:45 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2022/03/14 16:45:45 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2022/03/14 16:45:45 [lang:3] localOut: false:localAudio
2022/03/14 16:45:45 [source:3] Using ALSA 1.2.4.
2022/03/14 16:45:45 [source:3] Falling back on interleaved S16LE
2022/03/14 16:45:45 [source:3] Samplefreq=44100Hz, Bufsize=1764B, Frame=4B, Periods=4
2022/03/14 16:45:45 [lang:3] localOut: true:localAudio
2022/03/14 16:45:45 [sandbox:3] Sandboxing disabled
2022/03/14 16:45:45 [video.converter:3] Couldn't find preferred video converter: ffmpeg.
2022/03/14 16:45:45 [audio.converter:3] Using samplerate converter: native.
2022/03/14 16:45:45 [single_0:3] testSong.wav is static, resolving once for all...
2022/03/14 16:45:45 [clock.alsa:3] Streaming loop starts in auto-sync mode
2022/03/14 16:45:45 [clock.alsa:3] Delegating synchronisation to CPU clock
2022/03/14 16:45:46 [testSong.wav:3] Prepared "testSong.wav" (RID 0).
^C2022/03/14 16:46:11 [main:3] Shutdown started!
2022/03/14 16:46:11 [main:3] Waiting for main threads to terminate...
2022/03/14 16:46:11 [clock.alsa:3] Streaming loop stopped.
2022/03/14 16:46:11 [main:3] Main threads terminated.
2022/03/14 16:46:11 [threads:3] Shutting down scheduler...
2022/03/14 16:46:11 [threads:3] Scheduler shut down.
2022/03/14 16:46:11 [main:3] Cleaning downloaded files...
2022/03/14 16:46:11 [main:3] Freeing memory...
liquidsoap@pithree:~/audiobox $
